src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -4059,6 +4059,82 @@ void C2_MacroAssembler::masked_op(int ideal_opc, int mask_len, KRegister dst,
   }
 }
 
+/*
+ * Algorithm for vector D2L and F2I conversions:-
+ * a) Perform vector D2L/F2I cast.
+ * b) Choose fast path if none of the result vector lane contains 0x80000000 value.
+ *    It signifies that source value could be any of the special floating point
+ *    values(NaN,-Inf,Int,Max,-Min).
+ * c) Set destination to zero if source is NaN value.
+ * d) Replace 0x80000000 with MaxInt if source lane contains a +ve value.
+ */
+
+void C2_MacroAssembler::vector_castD2L_evex(XMMRegister dst, XMMRegister src, XMMRegister xtmp1, XMMRegister xtmp2,
+                                            KRegister ktmp1, AddressLiteral double_sign_flip,
+                                            AddressLiteral max_long, Register scratch, int vec_enc) {
+  Label done;
+  evcvttpd2qq(dst, src, vec_enc);
+  evmovdqul(xtmp1, k0, double_sign_flip, true, vec_enc, scratch);
+  evpcmpeqq(ktmp1, xtmp1, dst, vec_enc);
+  kortestwl(ktmp1, ktmp1);
+  jccb(Assembler::equal, done);
+
+  vpxor(xtmp2, xtmp2, xtmp2, vec_enc);
+  evcmppd(ktmp1, k0, src, src, Assembler::UNORD_Q, vec_enc);
+  evblendmpd(dst, ktmp1, dst, xtmp2, true, vec_enc);
+
+  evpcmpeqq(ktmp1, xtmp1, dst, vec_enc);
```

Comment:
Hi, some tiny suggestions here,

By using an extra `KRegister`, we can xor the previous masks to obtain this, eliminate the need of an `evpcmpeqq` here.

Reply:
Yes, nice suggestions,  AVX512 is a very powerful instruction set and opmask register allocation and our recent masking support contributions are major infrastructure enhancements which will enable developers to emit optimized AVX512 instruction sequence and harness its true potential.